### ISSUE TITLE
Add content for machine learning

### ISSUE THREAD

**Issue by Muhammad-Sarfaraz** [Muhammad-Sarfaraz/ByteLogs#5]

**Comment by KawsarAhmad43** [comment]
Add Data Pre-processing that might help others

**Comment by KawsarAhmad43** [comment]
#basic example of data pre-processing accessing titanic dataset
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = pd.read_csv(url)

# Drop irrelevant columns
data.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1, inplace=True)

# Handling missing values
data['Age'].fillna(data['Age'].median(), inplace=True)
data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)

# Encoding categorical variables
label_encoder = LabelEncoder()
data['Sex'] = label_encoder.fit_transform(data['Sex'])
data['Embarked'] = label_encoder.fit_transform(data['Embarked'])

# Feature scaling
scaler = StandardScaler()
data[['Age', 'Fare']] = scaler.fit_transform(data[['Age', 'Fare']])

# Splitting the dataset into train and test sets
X = data.drop('Survived', axis=1)
y = data['Survived']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Print the preprocessed data
print("Preprocessed data:")
print(X_train.head())
print(y_train.head())

**Comment by KawsarAhmad43** [comment]
# Correlation and Feature Reduction:

## add libraries
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import LabelEncoder

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = pd.read_csv(url)

# Drop irrelevant columns
data.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1, inplace=True)


# Handling missing values
data['Age'].fillna(data['Age'].median(), inplace=True)
data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)

# Encoding categorical variables
label_encoder = LabelEncoder()
data['Sex'] = label_encoder.fit_transform(data['Sex'])
data['Embarked'] = label_encoder.fit_transform(data['Embarked'])

# Encoding categorical variables
label_encoder = LabelEncoder()
data['Sex'] = label_encoder.fit_transform(data['Sex'])
data['Embarked'] = label_encoder.fit_transform(data['Embarked'])

# Calculate correlation matrix
corr_matrix = data.corr()
# Plot correlation matrix heatmap
plt.figure(figsize=(10, 8))
sns.heatmap(corr_matrix, annot=True, cmap="coolwarm")
plt.title("Correlation Matrix")
plt.show()

# Feature reduction using PCA
pca = PCA(n_components=2)
X_reduced = pca.fit_transform(data.drop('Survived', axis=1))

# Get the feature names
feature_names = data.drop('Survived', axis=1).columns

# Print the reduced features with names
print("Reduced features:")
for i in range(len(X_reduced)):
    print(f"Features: {list(zip(feature_names, X_reduced[i]))}")





#### Compiled Code with output: https://colab.research.google.com/drive/1wvO1iTqEh4mcgVBBoY_UHnmYubLJM3Sg?usp=sharing

**Comment by Muhammad-Sarfaraz** [comment]
What is the benefit, can you please elaborate?

**Comment by KawsarAhmad43** [comment]
Dear @Muhammad-Sarfaraz  Data pre-processing is essential when you are working with a dataset. a dataset can consist of null values, inappropriate structure that might create problem when implementing any ML algorithm. to prevent this we need data pre-processing.
